restapi/src/main/java/io/stargate/web/docsapi/service/write/DocumentWriteService.java

```diff
@@ -0,0 +1,224 @@
+      boolean numericBooleans,
+      ExecutionContext context) {
+    // create and cache the remove query prepare
+    Single<? extends Query<? extends BoundQuery>> deleteQueryPrepare =
+        prepareDeleteDocumentQuery(dataStore, keyspace, collection);
+
+    // create and cache the insert query prepare
+    Single<? extends Query<? extends BoundQuery>> insertQueryPrepare =
+        prepareInsertDocumentRowQuery(dataStore, keyspace, collection);
+
+    // execute when both done
+    return Single.zip(deleteQueryPrepare, insertQueryPrepare, Pair::of)
+        .observeOn(Schedulers.computation())
+        .map(
+            pair -> {
+              // take current timestamp for binding
+              long timestamp = timeSource.currentTimeMicros();
+
+              // create list of queries to send in batch
+              List<BoundQuery> queries = new ArrayList<>(rows.size() + 1);
+
+              // add delete
+              BoundQuery deleteExistingQuery =
+                  deleteQueryBuilder.bind(pair.getLeft(), documentId, timestamp - 1);
+              queries.add(deleteExistingQuery);
+
+              // then all inserts
+              rows.forEach(
+                  row -> {
+                    BoundQuery query =
+                        insertQueryBuilder.bind(
+                            pair.getRight(), documentId, row, timestamp, numericBooleans);
+                    queries.add(query);
+                  });
+
+              return queries;
+            })
+        // then execute batch
+        .flatMap(
+            boundQueries -> executeBatch(dataStore, boundQueries, context.nested("ASYNC UPDATE")))
+
+        // move away from the data store thread
+        .observeOn(Schedulers.io());
+  }
+
+  // create and cache the remove query prepare
+  private Single<? extends Query<? extends BoundQuery>> prepareInsertDocumentRowQuery(
+      DataStore dataStore, String keyspace, String collection) {
+
+    return Single.fromCallable(
+            () -> insertQueryBuilder.buildQuery(dataStore::queryBuilder, keyspace, collection))
+        .flatMap(query -> Single.fromCompletionStage(dataStore.prepare(query)))
+        .cache();
+  }
+
+  // create and cache the insert query prepare
+  private Single<? extends Query<? extends BoundQuery>> prepareDeleteDocumentQuery(
+      DataStore dataStore, String keyspace, String collection) {
+
+    return Single.fromCallable(
+            () -> deleteQueryBuilder.buildQuery(dataStore::queryBuilder, keyspace, collection))
+        .flatMap(query -> Single.fromCompletionStage(dataStore.prepare(query)))
+        .cache();
+  }
+
+  // executes batch
+  private SingleSource<ResultSet> executeBatch(
+      DataStore dataStore, List<BoundQuery> boundQueries, ExecutionContext context) {
+
+    // trace queries in context
+    boundQueries.forEach(context::traceDeferredDml);
+
+    // then execute batch
+    boolean loggedBatch = useLoggedBatches.orElse(dataStore.supportsLoggedBatches());
```

Comment:
Why not do this in the constructor so we don't need the `orElse` everywhere?

Reply:
Because the data store is supplied in the method call. The writer needs a pointer to `DataStore` the write should be executed on.